src/context.rs

```diff
@@ -0,0 +1,85 @@
+use indexmap::IndexSet;
+use std::convert::TryInto;
+use std::hash::Hash;
+
+/// Context object with global resources for working with grammar,
+/// such as interners.
+pub struct Context<Pat> {
+    interners: Interners<Pat>,
+}
+
+/// Dispatch helper, to allow implementing interning logic on
+/// the type passed to `cx.intern(...)`.
+pub trait InternInCx<Pat> {
+    type Interned;
+
+    fn intern_in_cx(self, cx: &mut Context<Pat>) -> Self::Interned;
+}
+
+impl<Pat> Context<Pat> {
+    pub fn new() -> Self {
+        Context {
+            interners: Interners::default(),
+        }
+    }
+
+    pub fn intern<T: InternInCx<Pat>>(&mut self, x: T) -> T::Interned {
+        x.intern_in_cx(self)
+    }
+}
+
+macro_rules! interners {
+    ($($name:ident => $ty:ty),* $(,)?) => {
+        #[allow(non_snake_case)]
+        struct Interners<Pat> {
+            $($name: IndexSet<$ty>),*
+        }
+
+        impl<Pat> Default for Interners<Pat> {
+            fn default() -> Self {
+                Interners {
+                    $($name: IndexSet::new()),*
+                }
+            }
+        }
+
+        $(
+            #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+            pub struct $name(u32);
+
+            impl<Pat: Eq + Hash> InternInCx<Pat> for $ty {
+                type Interned = $name;
+
+                fn intern_in_cx(self, cx: &mut Context<Pat>) -> Self::Interned {
+                    $name(cx.interners.$name.insert_full(self).0.try_into().unwrap())
+                }
+            }
+
+            impl<Pat> std::ops::Index<$name> for Context<Pat> {
+                type Output = $ty;
+
+                fn index(&self, interned: $name) -> &Self::Output {
+                    self.interners.$name.get_index(interned.0 as usize).unwrap()
+                }
+            }
+        )*
+    };
+}
+
+interners! {
+    IStr => String,
```

Comment:
Any reason for this being `String` and not `Box<str>`?

Reply:
Not really, I don't think the distinction matters much here (and `Box<str>` is "weirder").

Reply:
Personally, I prefer `Box<str>` for interners (it's two pointer sized instead of three!) but `String` is indeed the `ToOwned::Owned` for `str`, so I guess it's a tossup which to choose. (You need to use `Into<Box<str>>` instead of `ToOwned<str>` to use the box.)

Reply:
It doesn't really matter and using `ToOwned::Owned` makes sense.